APISyncExternalModule.php

```diff
@@ -898,25 +905,291 @@ function cancelSync(){
+					}
+				}
+			}
+		}
+	}
 
-		$result = $this->query("
-			select field_name
-			from redcap_metadata
-			where project_id = $pid
-			order by field_order
-			limit 1
-		");
+	private function translationsAreBuilt($project) {
+		$translation_settings = [
+			'form-translations',
+			'event-translations',
+			'export-form-translations',
+			'export-event-translations'
+		];
+		foreach($translation_settings as $name) {
+			if (gettype($project[$name]) == 'array') {
+				return true;
+			}
+		}
+	}
 
-		$row = $result->fetch_assoc();
+	private function getProjectTypePrefix(&$project) {
+		foreach ($project as $setting_name => $setting_value) {
+			if (strpos($setting_name, 'export-') !== false) {
+				return 'export-';
+			}
+		}
+		return '';
+	}
 
-		return $row['field_name'];
+	private function formatEventName($event_label) {
+		// // code copied from \Project::getUniqueEventNames
+		$event_name = trim(label_decode($event_label));
+		// Remove all spaces and non-alphanumeric characters, then make it lower case.
+		$event_name = preg_replace("/[^0-9a-z_ ]/i", '', $event_name);
+		$event_name = strtolower(substr(str_replace(" ", "_", $event_name), 0, 18));
+		// Remove any underscores at the beginning
+		while (substr($event_name, 0, 1) == "_") {
+			$event_name = substr($event_name, 1);
+		}
+		// Remove any underscores at the end
+		while (substr($event_name, -1, 1) == "_") {
+			$event_name = substr($event_name, 0, -1);
+		}
+		// If event name is still blank (maybe because of using multi-byte characters)
+		if ($event_name == '') {
+			// Get first 10 letters of MD5 of the event label
+			$event_name = substr(md5($event_label), 0, 10);
+		}
+		return $event_name;
+	}
+
+	private function formatFormName($form_label) {
+		// code copied from \REDCap::setFormName
+		$form_name = strip_tags(label_decode($form_label));
+		$form_name = preg_replace("/[^a-z0-9_]/", "", str_replace(" ", "_", strtolower(html_entity_decode($form_name, ENT_QUOTES))));
+		// Remove any double underscores, beginning numerals, and beginning/ending underscores
+		while (strpos($form_name, "__") !== false) 		$form_name = str_replace("__", "_", $form_name);
+		while (substr($form_name, 0, 1) == "_") 		$form_name = substr($form_name, 1);
+		while (substr($form_name, -1) == "_") 			$form_name = substr($form_name, 0, -1);
+		while (is_numeric(substr($form_name, 0, 1))) 	$form_name = substr($form_name, 1);
+		while (substr($form_name, 0, 1) == "_") 		$form_name = substr($form_name, 1);
+		// Cannot begin with numeral and cannot be blank
+		if (is_numeric(substr($form_name, 0, 1)) || $form_name == "") {
+			$form_name = substr(preg_replace("/[0-9]/", "", md5($form_name)), 0, 4) . $form_name;
+		}
+		// Make sure it's less than 50 characters long
+		$form_name = substr($form_name, 0, 50);
+		while (substr($form_name, -1) == "_") $form_name = substr($form_name, 0, -1);
+		return $form_name;
+	}
+
+	public function importTranslationsFile() {
```

Comment:
It looks like this and all the methods below here might not be used anymore.  Should they be removed?

Reply:
importTranslationsFile/Table, validateImport, and saveTranslations are all used for 'config_translations.php'. I've checked all the other methods I've added and deemed them necessary.

Reply:
Ah, my bad, that makes sense.